diff -Naurw cryptography-0.9.3/src/cryptography/hazmat/bindings/openssl/ec.py cryptography-0.9.3.new/src/cryptography/hazmat/bindings/openssl/ec.py
--- cryptography-0.9.3/src/cryptography/hazmat/bindings/openssl/ec.py	2015-07-09 07:41:12.000000000 -0700
+++ cryptography-0.9.3.new/src/cryptography/hazmat/bindings/openssl/ec.py	2015-08-08 12:15:42.739955819 -0700
@@ -390,7 +390,7 @@
 #endif
 
 #if defined(OPENSSL_NO_EC) || OPENSSL_VERSION_NUMBER < 0x1000200f || \
-    defined(LIBRESSL_VERSION_NUMBER)
+    defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x20020002L
 static const long Cryptography_HAS_EC_1_0_2 = 0;
 const char *(*EC_curve_nid2nist)(int) = NULL;
 #else
