diff -Naur FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/libfreerdp/crypto/opensslcompat.c FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/libfreerdp/crypto/opensslcompat.c
--- FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/libfreerdp/crypto/opensslcompat.c	2016-12-23 17:04:46.241267527 -0800
+++ FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/libfreerdp/crypto/opensslcompat.c	2016-12-23 17:04:54.787360289 -0800
@@ -19,7 +19,7 @@
 
 #include "opensslcompat.h"
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 
 BIO_METHOD* BIO_meth_new(int type, const char* name)
 {
diff -Naur FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/libfreerdp/crypto/opensslcompat.h FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/libfreerdp/crypto/opensslcompat.h
--- FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/libfreerdp/crypto/opensslcompat.h	2016-12-23 17:04:46.242267537 -0800
+++ FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/libfreerdp/crypto/opensslcompat.h	2016-12-23 17:05:12.651554194 -0800
@@ -30,7 +30,7 @@
 
 #include <openssl/opensslv.h>
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 
 #include <openssl/bio.h>
 #include <openssl/rsa.h>
diff -Naur FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/libfreerdp/crypto/tls.c FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/libfreerdp/crypto/tls.c
--- FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/libfreerdp/crypto/tls.c	2016-12-23 17:04:46.243267548 -0800
+++ FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/libfreerdp/crypto/tls.c	2016-12-23 17:04:54.789360311 -0800
@@ -323,7 +323,7 @@
 		case BIO_CTRL_PUSH:
 			if (next_bio && (next_bio != ssl_rbio))
 			{
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 				SSL_set_bio(tls->ssl, next_bio, next_bio);
 				CRYPTO_add(&(bio->next_bio->references), 1, CRYPTO_LOCK_BIO);
 #else
@@ -346,7 +346,7 @@
 				if (ssl_rbio != ssl_wbio)
 					BIO_free_all(ssl_wbio);
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 				if (next_bio)
 					CRYPTO_add(&(bio->next_bio->references), -1, CRYPTO_LOCK_BIO);
 				tls->ssl->wbio = tls->ssl->rbio = NULL;
@@ -384,7 +384,7 @@
 					BIO_push(ssl_rbio, next_bio);
 
 				BIO_set_next(bio, ssl_rbio);
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 				CRYPTO_add(&(ssl_rbio->references), 1, CRYPTO_LOCK_BIO);
 #else
 				BIO_up_ref(ssl_rbio);
@@ -1021,7 +1021,7 @@
  *        SSL struct is opaqe now
  */
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 	if (tls->alertDescription != TLS_ALERT_DESCRIPTION_CLOSE_NOTIFY)
 	{
 		/**
diff -Naur FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/winpr/libwinpr/crypto/hash.c FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/winpr/libwinpr/crypto/hash.c
--- FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/winpr/libwinpr/crypto/hash.c	2016-12-23 17:04:46.315268330 -0800
+++ FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/winpr/libwinpr/crypto/hash.c	2016-12-23 17:04:54.790360322 -0800
@@ -148,7 +148,7 @@
 
 #if defined(WITH_OPENSSL)
 	HMAC_CTX* hmac = NULL;
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
 	if (!(hmac = (HMAC_CTX*) calloc(1, sizeof(HMAC_CTX))))
 		return NULL;
 	HMAC_CTX_init(hmac);
@@ -269,7 +269,7 @@
 	HMAC_CTX* hmac = (HMAC_CTX*) ctx;
 	if (hmac)
 	{
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
 		HMAC_CTX_cleanup(hmac);
 		free(hmac);
 #else
@@ -319,7 +319,7 @@
 
 #if defined(WITH_OPENSSL)
 	EVP_MD_CTX* mdctx;
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
 	mdctx = EVP_MD_CTX_create();
 #else
 	mdctx = EVP_MD_CTX_new();
@@ -408,7 +408,7 @@
 	EVP_MD_CTX* mdctx = (EVP_MD_CTX*) ctx;
 	if (mdctx)
 	{
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
 		EVP_MD_CTX_destroy(mdctx);
 #else
 		EVP_MD_CTX_free(mdctx);
diff -Naur FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/winpr/libwinpr/utils/ssl.c FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/winpr/libwinpr/utils/ssl.c
--- FreeRDP-8d0809cf26868fc5ff651850a32724650597747f.orig/winpr/libwinpr/utils/ssl.c	2016-12-23 17:04:46.364268862 -0800
+++ FreeRDP-8d0809cf26868fc5ff651850a32724650597747f/winpr/libwinpr/utils/ssl.c	2016-12-23 17:04:54.790360322 -0800
@@ -44,7 +44,7 @@
  * set locking callbacks to use OpenSSL in a multi-threaded environment.
  */
 
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
 
 #define WINPR_OPENSSL_LOCKING_REQUIRED 1
 
