--- src/Csocket.cpp.orig	2015-08-04 08:00:04.000000000 +0000
+++ src/Csocket.cpp	2015-12-27 14:41:42.217596776 +0000
@@ -47,6 +47,7 @@
 #include <stdio.h>
 #include <openssl/conf.h>
 #include <openssl/engine.h>
+#include <openssl/comp.h>
 #endif /* HAVE_LIBSSL */
 
 #ifdef HAVE_ICU
@@ -572,6 +573,7 @@
 	}
 #endif /* _WIN32 */
 
+#ifndef OPENSSL_NO_COMP
 	COMP_METHOD *cm = NULL;
 
 	if( CT_ZLIB & eCompressionType )
@@ -587,6 +589,7 @@
 		if( cm )
 			SSL_COMP_add_compression_method( CT_RLE, cm );
 	}
+#endif
 
 	// setting this up once in the begining
 	s_iCsockSSLIdx = SSL_get_ex_new_index( 0, NULL, NULL, NULL, NULL );
@@ -1463,14 +1466,6 @@
 
 	switch( m_iMethod )
 	{
-	case SSL3:
-		m_ssl_ctx = SSL_CTX_new( SSLv3_client_method() );
-		if( !m_ssl_ctx )
-		{
-			CS_DEBUG( "WARNING: MakeConnection .... SSLv3_client_method failed!" );
-			return( false );
-		}
-		break;
 	case TLS12:
 #if defined( TLS1_2_VERSION ) && defined( OPENSSL_VERSION_NUMBER ) && OPENSSL_VERSION_NUMBER >= 0x1000100f
 		m_ssl_ctx = SSL_CTX_new( TLSv1_2_client_method() );
@@ -1510,6 +1505,17 @@
 		break;
 #endif /* OPENSSL_NO_SSL2 */
 		/* Fall through if SSL2 is disabled */
+#ifndef OPENSSL_NO_SSL3
+	case SSL3:
+		m_ssl_ctx = SSL_CTX_new( SSLv3_client_method() );
+		if( !m_ssl_ctx )
+		{
+			CS_DEBUG( "WARNING: MakeConnection .... SSLv3_client_method failed!" );
+			return( false );
+		}
+		break;
+#endif /* OPENSSL_NO_SSL3 */
+		/* Fall through if SSL3 is disabled */
 	case SSL23:
 	default:
 		if( m_iMethod != SSL23 )
@@ -1585,14 +1591,6 @@
 	SSL_CTX * pCTX = NULL;
 	switch( m_iMethod )
 	{
-	case SSL3:
-		pCTX = SSL_CTX_new( SSLv3_server_method() );
-		if( !pCTX )
-		{
-			CS_DEBUG( "WARNING: MakeConnection .... SSLv3_server_method failed!" );
-			return( NULL );
-		}
-		break;
 	case TLS12:
 #if defined( TLS1_2_VERSION ) && defined( OPENSSL_VERSION_NUMBER ) && OPENSSL_VERSION_NUMBER >= 0x1000100f
 		pCTX = SSL_CTX_new( TLSv1_2_server_method() );
@@ -1632,6 +1630,17 @@
 		break;
 #endif /* OPENSSL_NO_SSL2 */
 		/* Fall through if SSL2 is disabled */
+#ifndef OPENSSL_NO_SSL3
+	case SSL3:
+		pCTX = SSL_CTX_new( SSLv3_server_method() );
+		if( !pCTX )
+		{
+			CS_DEBUG( "WARNING: MakeConnection .... SSLv3_server_method failed!" );
+			return( NULL );
+		}
+		break;
+#endif /* OPENSSL_NO_SSL3 */
+		/* Fall through if SSL3 is disabled */
 	case SSL23:
 	default:
 		if( m_iMethod != SSL23 )
