diff -Naur links-2.8.orig/configure.in links-2.8/configure.in
--- links-2.8.orig/configure.in	2013-09-16 17:11:46.000000000 -0700
+++ links-2.8/configure.in	2014-12-21 15:29:11.706749792 -0800
@@ -779,7 +779,8 @@
 			AC_MSG_CHECKING([for OpenSSL])
 			CPPFLAGS="$CPPFLAGS_XX $OPENSSL_CFLAGS"
 			LIBS="$OPENSSL_LIBS $LIBS_XX"
-			AC_TRY_LINK([#include <openssl/ssl.h>], [SSLeay_add_ssl_algorithms()], cf_result=openssl, cf_result=no)
+			AC_TRY_LINK([#include <stdint.h>
+                         #include <openssl/ssl.h>], [SSLeay_add_ssl_algorithms()], cf_result=openssl, cf_result=no)
 			AC_MSG_RESULT($cf_result)
 		fi
 	fi
@@ -798,7 +799,8 @@
 					LIBS="-lssl -lcrypto $LIBS_XX"
 					CPPFLAGS="$CPPFLAGS_XX"
 				fi
-				AC_TRY_LINK([#include <openssl/ssl.h>], [SSLeay_add_ssl_algorithms()], cf_result=openssl, cf_result=no)
+				AC_TRY_LINK([#include <stdint.h>
+                             #include <openssl/ssl.h>], [SSLeay_add_ssl_algorithms()], cf_result=openssl, cf_result=no)
 			fi
 		done
 		AC_MSG_RESULT($cf_result)
diff -Naur links-2.8.orig/links.h links-2.8/links.h
--- links-2.8.orig/links.h	2013-09-16 17:11:46.000000000 -0700
+++ links-2.8/links.h	2014-12-21 15:19:12.070065258 -0800
@@ -193,6 +193,7 @@
 #if defined(OPENVMS) && defined(__VAX)
 #define OPENSSL_NO_SHA512
 #endif
+#include <stdint.h>
 #include <openssl/ssl.h>
 #include <openssl/rand.h>
 #endif
