diff -Naur ruby-2_0_0_648.orig/ext/openssl/extconf.rb ruby-2_0_0_648/ext/openssl/extconf.rb
--- ruby-2_0_0_648.orig/ext/openssl/extconf.rb	2015-12-16 04:31:18.000000000 -0800
+++ ruby-2_0_0_648/ext/openssl/extconf.rb	2016-03-28 13:32:05.837065312 -0700
@@ -103,6 +103,9 @@
 have_func("SSLv2_method")
 have_func("SSLv2_server_method")
 have_func("SSLv2_client_method")
+have_func("SSLv3_method")
+have_func("SSLv3_server_method")
+have_func("SSLv3_client_method")
 have_func("TLSv1_1_method")
 have_func("TLSv1_1_server_method")
 have_func("TLSv1_1_client_method")
diff -Naur ruby-2_0_0_648.orig/ext/openssl/ossl_ssl.c ruby-2_0_0_648/ext/openssl/ossl_ssl.c
--- ruby-2_0_0_648.orig/ext/openssl/ossl_ssl.c	2015-12-16 04:31:18.000000000 -0800
+++ ruby-2_0_0_648/ext/openssl/ossl_ssl.c	2016-03-28 13:32:05.839065329 -0700
@@ -129,9 +129,12 @@
     OSSL_SSL_METHOD_ENTRY(SSLv2_server),
     OSSL_SSL_METHOD_ENTRY(SSLv2_client),
 #endif
+#if defined(HAVE_SSLV3_METHOD) && defined(HAVE_SSLV3_SERVER_METHOD) && \
+        defined(HAVE_SSLV3_CLIENT_METHOD)
     OSSL_SSL_METHOD_ENTRY(SSLv3),
     OSSL_SSL_METHOD_ENTRY(SSLv3_server),
     OSSL_SSL_METHOD_ENTRY(SSLv3_client),
+#endif
     OSSL_SSL_METHOD_ENTRY(SSLv23),
     OSSL_SSL_METHOD_ENTRY(SSLv23_server),
     OSSL_SSL_METHOD_ENTRY(SSLv23_client),
