--- a/3rd_party/libsrp6a-sha512/t_math.c
+++ b/3rd_party/libsrp6a-sha512/t_math.c
@@ -720,7 +720,7 @@ BigIntegerModExp(BigInteger r, BigInteger b, BigInteger e, BigInteger m, BigInte
   else if(a == NULL) {
     BN_mod_exp(r, b, e, m, c);
   }
-#if OPENSSL_VERSION_NUMBER >= 0x00906000
+#if OPENSSL_VERSION_NUMBER >= 0x00906000 && !defined(LIBRESSL_VERSION_NUMBER)
   else if(B > 0 && B < ULONG_MAX) {  /* 0.9.6 and above has mont_word optimization */
     BN_mod_exp_mont_word(r, B, e, m, c, a);
   }
--- a/common/userpref.c
+++ b/common/userpref.c
@@ -627,7 +627,9 @@ userpref_error_t pair_record_generate_keys_and_certs(plist_t pair_record, key_da
 		}
 	}
 
+#ifndef LIBRESSL_VERSION_NUMBER
 	X509V3_EXT_cleanup();
+#endif
 	X509_free(dev_cert);
 
 	EVP_PKEY_free(pubkey);
