--- a/SSLeay.xs	2019-03-02 13:25:21.964959279 -0500
+++ b/SSLeay.xs	2019-03-02 13:37:01.684711228 -0500
@@ -5680,7 +5680,7 @@
 	const EVP_MD *h;
 	int md_size = -1;
 	c = s->enc_read_ctx->cipher;
-#if OPENSSL_VERSION_NUMBER >= 0x10001000L
+#if OPENSSL_VERSION_NUMBER >= 0x10001000L && !defined(LIBRESSL_VERSION_NUMBER)
 	if (s->s3)
 	    md_size = s->s3->tmp.new_mac_secret_size;
 #elif OPENSSL_VERSION_NUMBER >= 0x00909000L
