From: orbea <orbea@riseup.net>
Date: Fri, 8 Jul 2022 20:23:12 -0700
Subject: [PATCH] Support LibreSSL 3.5.x

diff --git a/libssh2-sys-0.2.23/build.rs b/libssh2-sys-0.2.23/build.rs
index d1d33dc..7f82663 100644
--- a/libssh2-sys-0.2.23/build.rs
+++ b/libssh2-sys-0.2.23/build.rs
@@ -121,6 +121,7 @@ fn main() {
         cfg.define("HAVE_EVP_AES_128_CTR", None);
         cfg.define("HAVE_POLL", None);
         cfg.define("HAVE_GETTIMEOFDAY", None);
+        cfg.define("HAVE_OPAQUE_STRUCTS", None);
 
         cfg.file("libssh2/src/openssl.c");
 
diff --git a/openssl-sys-0.9.72/build/main.rs b/openssl-sys-0.9.72/build/main.rs
index 243f146..d6d2eb4 100644
--- a/openssl-sys-0.9.72/build/main.rs
+++ b/openssl-sys-0.9.72/build/main.rs
@@ -254,6 +254,7 @@ See rust-openssl README for more information:
             (3, 3, _) => ('3', '3', 'x'),
             (3, 4, 0) => ('3', '4', '0'),
             (3, 4, _) => ('3', '4', 'x'),
+            (3, 5, _) => ('3', '5', 'x'),
             _ => version_error(),
         };
 
