--- a/source/helpers/pkcs7/pkcs7-openssl.c	2018-12-09 21:11:07.594403947 -0500
+++ b/source/helpers/pkcs7/pkcs7-openssl.c	2018-12-09 21:15:50.879934584 -0500
@@ -116,6 +116,11 @@
 #warning detect version of openssl at compile time
 #endif
 
+#ifdef LIBRESSL_VERSION_NUMBER
+#define PKCS12_SAFEBAG_get0_safes(bag) ((bag)->value.safes)
+#define PKCS12_SAFEBAG_get0_p8inf(bag) ((bag)->value.keybag)
+#endif
+
 typedef struct
 {
 	fz_context *ctx;
