diff -ur curl-7.39.0/lib/curl_ntlm_core.c new-curl/lib/curl_ntlm_core.c
--- curl-7.39.0/lib/curl_ntlm_core.c	2014-11-04 04:51:16.000000000 -0800
+++ new-curl/lib/curl_ntlm_core.c	2014-12-21 12:41:56.962952503 -0800
@@ -38,6 +38,7 @@
 #    ifndef OPENSSL_NO_MD4
 #      include <openssl/md4.h>
 #    endif
+#    include <stdint.h>
 #    include <openssl/md5.h>
 #    include <openssl/ssl.h>
 #    include <openssl/rand.h>
diff -ur curl-7.39.0/lib/urldata.h new-curl/lib/urldata.h
--- curl-7.39.0/lib/urldata.h	2014-11-04 23:23:42.000000000 -0800
+++ new-curl/lib/urldata.h	2014-12-21 12:35:56.946301391 -0800
@@ -81,6 +81,7 @@
 
 #ifdef USE_SSLEAY
 #ifdef USE_OPENSSL
+#include <stdint.h>
 #include <openssl/rsa.h>
 #include <openssl/crypto.h>
 #include <openssl/x509.h>
