--- syslog-ng-3.4.1.orig/configure.in
+++ syslog-ng-3.4.1/configure.in
@@ -68,7 +68,7 @@
 pidfiledir='${localstatedir}'
 moduledir='${exec_prefix}/lib/syslog-ng'
 
-AM_CONFIG_HEADER(config.h)
+AC_CONFIG_HEADERS(config.h)
 
 dnl ***************************************************************************
 dnl Arguments
@@ -227,7 +227,6 @@
 dnl ***************************************************************************
 dnl Checks for programs.
 AC_PROG_CC
-AM_PROG_CC_STDC
 AC_PROG_CC_C99
 if test "x$ac_cv_prog_cc_c99" = "xno"; then
 	AC_MSG_ERROR("C99 standard compliant C compiler required. Try GCC 3.x or later.")
--- syslog-ng-3.4.1.orig/modules/afmongodb/libmongo-client/configure.ac
+++ syslog-ng-3.4.1/modules/afmongodb/libmongo-client/configure.ac
@@ -15,7 +15,7 @@
 
 AC_CONFIG_MACRO_DIR([m4])
 m4_ifdef([AM_SILENT_RULES], [AM_SILENT_RULES([yes])])
-AM_CONFIG_HEADER(config.h)
+AC_CONFIG_HEADERS(config.h)
 
 AC_ARG_WITH([versioned-symbols], AC_HELP_STRING([--with-versioned-symbols],[Use versioned symbols]),[dnl
 vsymldflags="-Wl,--version-script,\${srcdir}/libmongo-client.ver -Wl,-O1"
@@ -24,7 +24,6 @@
 dnl ***************************************************************************
 dnl Checks for programs.
 AC_PROG_CC
-AM_PROG_CC_STDC
 AC_PROG_MAKE_SET
 PKG_PROG_PKG_CONFIG
 LT_INIT([shared])
