diff -Naur libressl-2.2.0.orig/crypto/compat/getentropy_linux.c libressl-2.2.0/crypto/compat/getentropy_linux.c
--- libressl-2.2.0.orig/crypto/compat/getentropy_linux.c	2015-06-30 23:42:09.873389199 -0700
+++ libressl-2.2.0/crypto/compat/getentropy_linux.c	2015-06-30 23:43:20.530057972 -0700
@@ -28,7 +28,6 @@
 #include <sys/resource.h>
 #include <sys/syscall.h>
 #ifdef SYS__sysctl
-#include <sys/sysctl.h>
 #include <linux/sysctl.h>
 #endif
 #include <sys/statvfs.h>
