diff -Naur libressl-2.1.2.orig/crypto/aes/aes-elf-x86_64.s libressl-2.1.2/crypto/aes/aes-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/aes/aes-elf-x86_64.s	2014-12-07 19:16:01.000000000 -0500
+++ libressl-2.1.2/crypto/aes/aes-elf-x86_64.s	2014-12-21 11:30:49.426981798 -0500
@@ -2539,3 +2539,7 @@
 .long	0x1b1b1b1b, 0x1b1b1b1b, 0, 0
 .byte	65,69,83,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	64
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/aes/aesni-elf-x86_64.s libressl-2.1.2/crypto/aes/aesni-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/aes/aesni-elf-x86_64.s	2014-12-07 19:16:02.000000000 -0500
+++ libressl-2.1.2/crypto/aes/aesni-elf-x86_64.s	2014-12-21 11:31:00.208982174 -0500
@@ -2533,3 +2533,7 @@
 
 .byte	65,69,83,32,102,111,114,32,73,110,116,101,108,32,65,69,83,45,78,73,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	64
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/aes/aesni-sha1-elf-x86_64.s libressl-2.1.2/crypto/aes/aesni-sha1-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/aes/aesni-sha1-elf-x86_64.s	2014-12-07 19:16:02.000000000 -0500
+++ libressl-2.1.2/crypto/aes/aesni-sha1-elf-x86_64.s	2014-12-21 11:30:53.942981955 -0500
@@ -1394,3 +1394,7 @@
 
 .byte	65,69,83,78,73,45,67,66,67,43,83,72,65,49,32,115,116,105,116,99,104,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	64
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/aes/bsaes-elf-x86_64.s libressl-2.1.2/crypto/aes/bsaes-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/aes/bsaes-elf-x86_64.s	2014-12-07 19:16:02.000000000 -0500
+++ libressl-2.1.2/crypto/aes/bsaes-elf-x86_64.s	2014-12-21 11:30:42.643981562 -0500
@@ -2496,3 +2496,7 @@
 .byte	66,105,116,45,115,108,105,99,101,100,32,65,69,83,32,102,111,114,32,120,56,54,95,54,52,47,83,83,83,69,51,44,32,69,109,105,108,105,97,32,75,195,164,115,112,101,114,44,32,80,101,116,101,114,32,83,99,104,119,97,98,101,44,32,65,110,100,121,32,80,111,108,121,97,107,111,118,0
 .align	64
 .size	_bsaes_const,.-_bsaes_const
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/aes/vpaes-elf-x86_64.s libressl-2.1.2/crypto/aes/vpaes-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/aes/vpaes-elf-x86_64.s	2014-12-07 19:16:01.000000000 -0500
+++ libressl-2.1.2/crypto/aes/vpaes-elf-x86_64.s	2014-12-21 11:31:10.445982530 -0500
@@ -826,3 +826,7 @@
 .byte	86,101,99,116,111,114,32,80,101,114,109,117,116,97,116,105,111,110,32,65,69,83,32,102,111,114,32,120,56,54,95,54,52,47,83,83,83,69,51,44,32,77,105,107,101,32,72,97,109,98,117,114,103,32,40,83,116,97,110,102,111,114,100,32,85,110,105,118,101,114,115,105,116,121,41,0
 .align	64
 .size	_vpaes_consts,.-_vpaes_consts
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/bn/gf2m-elf-x86_64.s libressl-2.1.2/crypto/bn/gf2m-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/bn/gf2m-elf-x86_64.s	2014-12-07 19:16:02.000000000 -0500
+++ libressl-2.1.2/crypto/bn/gf2m-elf-x86_64.s	2014-12-21 11:29:31.936979099 -0500
@@ -289,3 +289,7 @@
 .size	bn_GF2m_mul_2x2,.-bn_GF2m_mul_2x2
 .byte	71,70,40,50,94,109,41,32,77,117,108,116,105,112,108,105,99,97,116,105,111,110,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	16
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/bn/modexp512-elf-x86_64.s libressl-2.1.2/crypto/bn/modexp512-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/bn/modexp512-elf-x86_64.s	2014-12-07 19:16:02.000000000 -0500
+++ libressl-2.1.2/crypto/bn/modexp512-elf-x86_64.s	2014-12-21 11:29:42.082979452 -0500
@@ -1771,3 +1771,7 @@
 .Lepilogue:
 	.byte	0xf3,0xc3
 .size	mod_exp_512, . - mod_exp_512
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/bn/mont5-elf-x86_64.s libressl-2.1.2/crypto/bn/mont5-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/bn/mont5-elf-x86_64.s	2014-12-07 19:16:02.000000000 -0500
+++ libressl-2.1.2/crypto/bn/mont5-elf-x86_64.s	2014-12-21 11:29:48.583979679 -0500
@@ -782,3 +782,7 @@
 .long	0,0, 0,0, 0,0, -1,-1
 .long	0,0, 0,0, 0,0,  0,0
 .byte	77,111,110,116,103,111,109,101,114,121,32,77,117,108,116,105,112,108,105,99,97,116,105,111,110,32,119,105,116,104,32,115,99,97,116,116,101,114,47,103,97,116,104,101,114,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/bn/mont-elf-x86_64.s libressl-2.1.2/crypto/bn/mont-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/bn/mont-elf-x86_64.s	2014-12-07 19:16:02.000000000 -0500
+++ libressl-2.1.2/crypto/bn/mont-elf-x86_64.s	2014-12-21 11:29:56.422979952 -0500
@@ -1372,3 +1372,7 @@
 .size	bn_sqr4x_mont,.-bn_sqr4x_mont
 .byte	77,111,110,116,103,111,109,101,114,121,32,77,117,108,116,105,112,108,105,99,97,116,105,111,110,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	16
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/camellia/cmll-elf-x86_64.s libressl-2.1.2/crypto/camellia/cmll-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/camellia/cmll-elf-x86_64.s	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/camellia/cmll-elf-x86_64.s	2014-12-21 11:30:38.326981411 -0500
@@ -1836,3 +1836,7 @@
 .size	Camellia_cbc_encrypt,.-Camellia_cbc_encrypt
 
 .byte	67,97,109,101,108,108,105,97,32,102,111,114,32,120,56,54,95,54,52,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/cpuid-elf-x86_64.S libressl-2.1.2/crypto/cpuid-elf-x86_64.S
--- libressl-2.1.2.orig/crypto/cpuid-elf-x86_64.S	2014-12-07 19:16:04.000000000 -0500
+++ libressl-2.1.2/crypto/cpuid-elf-x86_64.S	2014-12-21 11:31:16.752982750 -0500
@@ -173,3 +173,7 @@
 	leaq	8(%rsp),%rax
 	.byte	0xf3,0xc3
 .size	OPENSSL_wipe_cpu,.-OPENSSL_wipe_cpu
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/Makefile.am.elf-x86_64 libressl-2.1.2/crypto/Makefile.am.elf-x86_64
--- libressl-2.1.2.orig/crypto/Makefile.am.elf-x86_64	2014-12-07 18:45:27.000000000 -0500
+++ libressl-2.1.2/crypto/Makefile.am.elf-x86_64	2014-12-21 11:30:26.586981002 -0500
@@ -39,3 +39,7 @@
 libcrypto_la_CFLAGS += -DOPENSSL_CPUID_OBJ
 libcrypto_la_SOURCES += $(ASM_X86_64_ELF)
 endif
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/md5/md5-elf-x86_64.s libressl-2.1.2/crypto/md5/md5-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/md5/md5-elf-x86_64.s	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/md5/md5-elf-x86_64.s	2014-12-21 11:30:02.071980149 -0500
@@ -666,3 +666,7 @@
 .Lepilogue:
 	.byte	0xf3,0xc3
 .size	md5_block_asm_data_order,.-md5_block_asm_data_order
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/modes/ghash-elf-x86_64.s libressl-2.1.2/crypto/modes/ghash-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/modes/ghash-elf-x86_64.s	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/modes/ghash-elf-x86_64.s	2014-12-21 11:31:32.516983299 -0500
@@ -1024,3 +1024,7 @@
 
 .byte	71,72,65,83,72,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	64
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/rc4/rc4-elf-x86_64.s libressl-2.1.2/crypto/rc4/rc4-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/rc4/rc4-elf-x86_64.s	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/rc4/rc4-elf-x86_64.s	2014-12-21 11:31:27.609983128 -0500
@@ -613,3 +613,7 @@
 .byte	82,67,52,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	64
 .size	RC4_options,.-RC4_options
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/rc4/rc4-md5-elf-x86_64.s libressl-2.1.2/crypto/rc4/rc4-md5-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/rc4/rc4-md5-elf-x86_64.s	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/rc4/rc4-md5-elf-x86_64.s	2014-12-21 11:31:22.634982955 -0500
@@ -1257,3 +1257,7 @@
 .Labort:
 	.byte	0xf3,0xc3
 .size	rc4_md5_enc,.-rc4_md5_enc
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/sha/sha1-elf-x86_64.s libressl-2.1.2/crypto/sha/sha1-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/sha/sha1-elf-x86_64.s	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/sha/sha1-elf-x86_64.s	2014-12-21 11:30:16.080980637 -0500
@@ -2484,3 +2484,7 @@
 .long	0x00010203,0x04050607,0x08090a0b,0x0c0d0e0f	
 .byte	83,72,65,49,32,98,108,111,99,107,32,116,114,97,110,115,102,111,114,109,32,102,111,114,32,120,56,54,95,54,52,44,32,67,82,89,80,84,79,71,65,77,83,32,98,121,32,60,97,112,112,114,111,64,111,112,101,110,115,115,108,46,111,114,103,62,0
 .align	64
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/sha/sha256-elf-x86_64.S libressl-2.1.2/crypto/sha/sha256-elf-x86_64.S
--- libressl-2.1.2.orig/crypto/sha/sha256-elf-x86_64.S	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/sha/sha256-elf-x86_64.S	2014-12-21 11:30:10.906980456 -0500
@@ -1776,3 +1776,7 @@
 .long	0x391c0cb3,0x4ed8aa4a,0x5b9cca4f,0x682e6ff3
 .long	0x748f82ee,0x78a5636f,0x84c87814,0x8cc70208
 .long	0x90befffa,0xa4506ceb,0xbef9a3f7,0xc67178f2
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/sha/sha512-elf-x86_64.S libressl-2.1.2/crypto/sha/sha512-elf-x86_64.S
--- libressl-2.1.2.orig/crypto/sha/sha512-elf-x86_64.S	2014-12-07 19:16:03.000000000 -0500
+++ libressl-2.1.2/crypto/sha/sha512-elf-x86_64.S	2014-12-21 11:30:21.762980834 -0500
@@ -1800,3 +1800,7 @@
 .quad	0x3c9ebe0a15c9bebc,0x431d67c49c100d4c
 .quad	0x4cc5d4becb3e42b6,0x597f299cfc657e2a
 .quad	0x5fcb6fab3ad6faec,0x6c44198c4a475817
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
diff -Naur libressl-2.1.2.orig/crypto/whrlpool/wp-elf-x86_64.s libressl-2.1.2/crypto/whrlpool/wp-elf-x86_64.s
--- libressl-2.1.2.orig/crypto/whrlpool/wp-elf-x86_64.s	2014-12-07 19:16:04.000000000 -0500
+++ libressl-2.1.2/crypto/whrlpool/wp-elf-x86_64.s	2014-12-21 11:30:31.768981183 -0500
@@ -856,3 +856,7 @@
 .byte	228,39,65,139,167,125,149,216
 .byte	251,238,124,102,221,23,71,158
 .byte	202,45,191,7,173,90,131,51
+
+#if defined(__linux__) && defined(__ELF__)
+.section .note.GNU-stack,"",%progbits
+#endif
