diff -Naurw libressl-2.3.0.orig/apps/Makefile.in libressl-2.3.0/apps/Makefile.in
--- libressl-2.3.0.orig/apps/Makefile.in	2015-09-22 05:06:38.000000000 -0700
+++ libressl-2.3.0/apps/Makefile.in	2015-09-24 12:59:38.106937861 -0700
@@ -310,7 +310,7 @@
 top_srcdir = @top_srcdir@
 AM_CFLAGS = -I$(top_srcdir)/include -I$(top_srcdir)/include/compat
 AM_CPPFLAGS = -DLIBRESSL_INTERNAL
-SUBDIRS = openssl nc
+SUBDIRS = openssl
 EXTRA_DIST = CMakeLists.txt
 all: all-recursive
 
