diff -Naur src.orig/context.c src/context.c
--- src.orig/context.c	2014-01-30 05:43:33.000000000 +0900
+++ src/context.c	2015-10-02 18:31:55.418277795 +0900
@@ -66,7 +66,9 @@
 static LSEC_SSL_METHOD* str2method(const char *method)
 {
   if (!strcmp(method, "sslv23"))  return SSLv23_method();
+#ifndef OPENSSL_NO_SSL3
   if (!strcmp(method, "sslv3"))   return SSLv3_method();
+#endif
   if (!strcmp(method, "tlsv1"))   return TLSv1_method();
 #if (OPENSSL_VERSION_NUMBER >= 0x1000100fL)
   if (!strcmp(method, "tlsv1_1")) return TLSv1_1_method();
