diff -Naurw M2Crypto-0.23.0.orig/SWIG/_ssl.i M2Crypto-0.23.0/SWIG/_ssl.i
--- M2Crypto-0.23.0.orig/SWIG/_ssl.i	2015-12-04 04:27:05.000000000 -0800
+++ M2Crypto-0.23.0/SWIG/_ssl.i	2016-02-18 16:46:48.725288975 -0800
@@ -56,8 +56,10 @@
 %rename(sslv2_method) SSLv2_method;
 extern SSL_METHOD *SSLv2_method(void);
 #endif
+#ifndef OPENSSL_NO_SSL3
 %rename(sslv3_method) SSLv3_method;
 extern SSL_METHOD *SSLv3_method(void);
+#endif
 %rename(sslv23_method) SSLv23_method;
 extern SSL_METHOD *SSLv23_method(void);
 %rename(tlsv1_method) TLSv1_method;
