diff -Naur sqlcipher-3.4.1.orig/src/crypto_openssl.c sqlcipher-3.4.1/src/crypto_openssl.c
--- sqlcipher-3.4.1.orig/src/crypto_openssl.c	2016-12-28 09:39:09.000000000 -0800
+++ sqlcipher-3.4.1/src/crypto_openssl.c	2017-06-17 15:43:13.970281603 -0700
@@ -46,7 +46,7 @@
 static unsigned int openssl_init_count = 0;
 static sqlite3_mutex* openssl_rand_mutex = NULL;
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 static HMAC_CTX *HMAC_CTX_new(void)
 {
   HMAC_CTX *ctx = OPENSSL_malloc(sizeof(*ctx));
