--- spamc/libspamc.c.orig	2015-04-29 04:56:59.000000000 +0000
+++ spamc/libspamc.c	2015-12-26 20:08:43.478248146 +0000
@@ -1216,7 +1216,7 @@
 	if (flags & SPAMC_TLSV1) {
 	    meth = TLSv1_client_method();
 	} else {
-	    meth = SSLv3_client_method(); /* default */
+	    meth = SSLv23_client_method(); /* default */
 	}
 	SSL_load_error_strings();
 	ctx = SSL_CTX_new(meth);
@@ -1604,7 +1604,7 @@
     if (flags & SPAMC_USE_SSL) {
 #ifdef SPAMC_SSL
 	SSLeay_add_ssl_algorithms();
-	meth = SSLv3_client_method();
+	meth = SSLv23_client_method();
 	SSL_load_error_strings();
 	ctx = SSL_CTX_new(meth);
 #else
