diff -Naur hostapd-2.6.orig/src/crypto/crypto_openssl.c hostapd-2.6/src/crypto/crypto_openssl.c
--- hostapd-2.6.orig/src/crypto/crypto_openssl.c	2017-02-28 12:49:32.286902599 -0800
+++ hostapd-2.6/src/crypto/crypto_openssl.c	2017-02-28 12:53:03.119653570 -0800
@@ -611,7 +611,7 @@
 
 void * dh5_init(struct wpabuf **priv, struct wpabuf **publ)
 {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 	DH *dh;
 	struct wpabuf *pubkey = NULL, *privkey = NULL;
 	size_t publen, privlen;
@@ -712,7 +712,7 @@
 
 void * dh5_init_fixed(const struct wpabuf *priv, const struct wpabuf *publ)
 {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 	DH *dh;
 
 	dh = DH_new();
